File: services/constraint-engine/content.py
import os
import yaml

_archetype_cache: dict[str, dict] = {}


def content_root() -> str:
    if v := os.getenv("CONTENT_DIR"):
        return os.path.join(v, "constraints")
    candidates = [
        os.path.join("..", "..", "content", "constraints"),
        os.path.join("content", "constraints"),
        "/app/content/constraints",
    ]
    for c in candidates:
        if os.path.isdir(c):
            return c
    return candidates[0]


def load_archetype(slug: str) -> dict:
    if slug in _archetype_cache:
        return _archetype_cache[slug]
    path = os.path.join(content_root(), slug, "levels.yaml")
    with open(path) as f:
        data = yaml.safe_load(f)
    if not data.get("slug"):
        data["slug"] = slug
    _archetype_cache[slug] = data
    return data


def list_archetypes() -> list[dict]:
    root = content_root()
    out = []
    for entry in os.listdir(root):
        entry_path = os.path.join(root, entry)
        if not os.path.isdir(entry_path):
            continue
        try:
            a = load_archetype(entry)
            out.append({
                "slug": a["slug"],
                "name": a["name"],
                "starting_monolith": a["starting_monolith"],
            })
        except Exception:
            pass
    return out


def origin_code(a: dict) -> str:
    oc = a.get("origin_code", "")
    if not oc:
        return ""
    path = os.path.join(content_root(), oc)
    try:
        with open(path) as f:
            return f.read()
    except Exception:
        return ""

File: services/constraint-engine/content.go
package main

import (
	"os"
	"path/filepath"
	"sync"

	"gopkg.in/yaml.v3"
)

var (
	archetypeCache = map[string]*Archetype{}
	cacheMu        sync.Mutex
)

// contentRoot resolves <repo>/content/constraints regardless of CWD.
func contentRoot() string {
	if v := os.Getenv("CONTENT_DIR"); v != "" {
		return filepath.Join(v, "constraints")
	}
	// Running from services/constraint-engine -> ../../content/constraints
	candidates := []string{
		filepath.Join("..", "..", "content", "constraints"),
		filepath.Join("content", "constraints"),
		"/app/content/constraints",
	}
	for _, c := range candidates {
		if st, err := os.Stat(c); err == nil && st.IsDir() {
			return c
		}
	}
	return candidates[0]
}

func loadArchetype(slug string) (*Archetype, error) {
	cacheMu.Lock()
	defer cacheMu.Unlock()
	if a, ok := archetypeCache[slug]; ok {
		return a, nil
	}
	path := filepath.Join(contentRoot(), slug, "levels.yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var a Archetype
	if err := yaml.Unmarshal(data, &a); err != nil {
		return nil, err
	}
	if a.Slug == "" {
		a.Slug = slug
	}
	archetypeCache[slug] = &a
	return &a, nil
}

func listArchetypes() []map[string]string {
	root := contentRoot()
	entries, _ := os.ReadDir(root)
	out := []map[string]string{}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		if a, err := loadArchetype(e.Name()); err == nil {
			out = append(out, map[string]string{
				"slug": a.Slug, "name": a.Name, "starting_monolith": a.StartingMonolith,
			})
		}
	}
	return out
}

func originCode(a *Archetype) string {
	if a.OriginCode == "" {
		return ""
	}
	path := filepath.Join(contentRoot(), a.OriginCode)
	if data, err := os.ReadFile(path); err == nil {
		return string(data)
	}
	return ""
}
